### ISSUE TITLE
Runs stuck as pending when workspace is deleted

### ISSUE THREAD

**Issue by Spritekin** [leg100/otf#577]
Hi @leg100 
I was developing my application, creating and destroying environments using OTF. Suddenly I noticed the API started returning run IDs however they stayed in a pending state.

So I went into the UI and found this:
<img width="893" alt="Screenshot 2023-08-27 at 9 03 48 pm" src="https://github.com/leg100/otf/assets/13467740/e3a6bc6d-9bbf-4145-8002-c4076c97706d">

Apparently all my runs were in a pending state, however there is nothing running.
So I went to the OTF logs and found this at around the same time:

```
2023/08/27 08:19:20 INFO enqueued plan id=run-YI30gZo8Ue2PvhjZ subject=scheduler
2023/08/27 08:19:20 INFO started plan id=run-YI30gZo8Ue2PvhjZ subject=local-agent
2023/08/27 08:19:20 INFO executing phase component=agent run=run-YI30gZo8Ue2PvhjZ phase=plan
2023/08/27 08:20:59 INFO finishing phase component=agent run=run-YI30gZo8Ue2PvhjZ phase=plan
2023/08/27 08:20:59 INFO finished plan id=run-YI30gZo8Ue2PvhjZ resource_changes=+0/~0/−1 output_changes=+0/~0/−0 subject=local-agent run_status=apply_queued
2023/08/27 08:20:59 INFO started apply id=run-YI30gZo8Ue2PvhjZ subject=local-agent
2023/08/27 08:20:59 INFO executing phase component=agent run=run-YI30gZo8Ue2PvhjZ phase=apply
2023/08/27 08:21:53 INFO created state version id=sv-5XpnvnoOCFTUh7G4 workspace=ws-xcCkz9gRRQw8nvno serial=13 subject=local-agent
2023/08/27 08:21:53 INFO finishing phase component=agent run=run-YI30gZo8Ue2PvhjZ phase=apply
2023/08/27 08:21:53 INFO finished apply id=run-YI30gZo8Ue2PvhjZ resource_changes=+0/~0/−1 output_changes=+0/~0/−0 subject=local-agent run_status=applied
2023/08/27 08:23:25 INFO disconnected workspace workspace=ws-xcCkz9gRRQw8nvno subject=Spritekin
2023/08/27 08:23:25 INFO deleted workspace id=ws-xcCkz9gRRQw8nvno name=oss-user-4 subject=Spritekin
2023/08/27 08:23:25 ERROR unsubscribing full subscriber component=broker error=<nil> sub=purger-vat4 queue_length=100
2023/08/27 08:23:25 ERROR unsubscribing full subscriber component=broker error=<nil> sub=scheduler-yW0c queue_length=100
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-YI30gZo8Ue2PvhjZ event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-CZuhkyrbQOMQzwEO event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-DtGtgNxsaPV7id9x event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-vxvKfB1ZShYSxQid event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-yb64Ug4FQzyYdRT5 event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-sNBd4YAIdfFygXxX event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-kzGOyCDeewJTXvKu event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-jmcZ4a9W50XJWjxJ event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-E99XmcDQmMIkXioM event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-I8Tym0tMiOkVce1m event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-Vgp8wpRgxBOnuT82 event=deleted
2023/08/27 08:23:25 ERROR unsubscribing full subscriber component=broker error=<nil> sub=chunk-proxy-WmkW queue_length=100
2023/08/27 08:23:25 ERROR unsubscribing full subscriber component=broker error=<nil> sub=reporter-YPrE queue_length=100
2023/08/27 08:23:25 ERROR unsubscribing full subscriber component=broker error=<nil> sub=run-watch-AqWk queue_length=100
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-vEeBG01uxhA348n0 event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-A8TBGiTDNtKC0DKq event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-CQDDjnaKFUmNnow3 event=deleted
2023/08/27 08:23:25 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-moS7jsOqQu0CKEKm event=deleted
2023/08/27 08:23:25 ERROR unsubscribing full subscriber component=broker error=<nil> sub=notifier-nAMQ queue_length=100
2023/08/27 08:23:25 ERROR deleting webhook error="DELETE https://api.github.com/repos/spritekindev/oss-user-4/hooks/430618832: 404 Not Found []" repo=spritekindev/oss-user-4 cloud=github
2023/08/27 08:25:43 INFO created webhook webhook.id=79bb67f5-7286-4789-9525-8ef50b856de0 webhook.vcs_provider_id=vcs-C05aynPJVWi9MdC4 webhook.cloud=github webhook.repo=spritekindev/oss-user-4 webhook.endpoint=https://otf.oss.ossy.dev/webhooks/vcs/79bb67f5-7286-4789-9525-8ef50b856de0
2023/08/27 08:25:43 INFO connected workspace repo workspace=ws-dOfx4AnUK1wR4qap subject=Spritekin repo=spritekindev/oss-user-4
2023/08/27 08:25:43 INFO created workspace id=ws-dOfx4AnUK1wR4qap name=oss-user-4 organization=Spritekin subject=Spritekin
2023/08/27 10:48:04 INFO created webhook webhook.id=890afe79-d339-4750-a71e-62aefc08c7d1 webhook.vcs_provider_id=vcs-C05aynPJVWi9MdC4 webhook.cloud=github webhook.repo=spritekindev/oss-user-5 webhook.endpoint=https://otf.oss.ossy.dev/webhooks/vcs/890afe79-d339-4750-a71e-62aefc08c7d1
2023/08/27 10:48:04 INFO connected workspace repo workspace=ws-IThr8a5hXtGgVieQ subject=Spritekin repo=spritekindev/oss-user-5
2023/08/27 10:48:04 INFO created workspace id=ws-IThr8a5hXtGgVieQ name=oss-user-5 organization=Spritekin subject=Spritekin
```
Let me explain.

Notice in the logs, at 08:19 I did a run and apply, in fact it was a destroy run because I wanted to delete all user resources, then I did run a workspace delete which can be observed at 08:23:25, however after that you may see some errors appearing.

Now after deleting the workspace oss-user-4 the next thing I did was to create a new workspace with the same name oss-user-4 which you can see at 08:25:43

But after that the new workspace wouldn't start any runs. It does return a success when a run is created and returns the new run ID but the runs are marked as pending and they wouldn't progress.

So I thought maybe its a problem because I used the same workspace name, so I created a new workspace oss-user-5 and tried more runs which you can see in the first screenshot. Runs just did not proceed.

My suspicion in this case is that there was some pending run in the queue when I decided to delete my workspace oss-user-4, with the workspace gone there was a problem reconciling the runs with the missing workspace which fired the "ERROR workspace queue does not exist for run event" errors. Probably a run just got stuck there.

After that we got more runs requests but given the run was blocked OTF just wouldn't process runs anymore.

Finally, I went into kubernetes and deleted the pod then the runs started processing.

So my suggestion, if a workspace is deleted, delete all related runs from the queue.

**Comment by Spritekin** [comment]
Still running more tests on workspace creation and destruction...

**Comment by Spritekin** [comment]
Ok so I did more tests, this is the pattern.

a. My Repositories are backed by VCS.
b. I run a destroy run so I remove all resources.
c. I delete the repository (via UI or via API same difference)
d. When I try to start a new run (by merge code, API call or UI) the run is placed in a planning state and never really starts.
e. Restarting the container unblocks all and the runs start.

**Comment by Spritekin** [comment]
@leg100 

Definitely an issue. Tested many times with the same result. This is a new example, from the OTF logs:
```
2023/09/01 12:22:57 INFO deleted workspace id=ws-a4sD6zhVJieCNpGd name=oss-user-6-oss-prod-changedetection-35 subject=Spritekin
2023/09/01 12:22:57 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-GgYSkVE75WkQqs81 event=deleted
2023/09/01 12:22:57 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-91NbqIiaImdGSy71 event=deleted
2023/09/01 12:22:57 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-7jRP4pWMO6PwBUCa event=deleted
2023/09/01 12:22:57 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-kaBFIjfEBdkQ6ddS event=deleted
2023/09/01 12:22:57 ERROR workspace queue does not exist for run event component=scheduler error=<nil> workspace="" run=run-cZoKskkNftDiKAkp event=deleted
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=chunk-proxy-nc8K queue_length=100
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=purger-MgaO queue_length=100
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=notifier-F1iB queue_length=100
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=run-watch-kb4P queue_length=100
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=scheduler-d9jm queue_length=100
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=reporter-48e8 queue_length=100
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=run-watch-5vwS queue_length=100
2023/09/01 12:29:19 INFO updated webhook webhook.id=ca33e812-4b94-4269-938a-a7e738edad9a webhook.vcs_provider_id=vcs-C05aynPJVWi9MdC4 webhook.cloud=github webhook.repo=spritekindev/oss-user-6 webhook.endpoint=https://otf.oss.ossy.dev/webhooks/vcs/ca33e812-4b94-4269-938a-a7e738edad9a webhook.vcs_id=430744892
2023/09/01 12:29:19 INFO connected workspace repo workspace=ws-ipRXmQVDr94eJQ1r subject=Spritekin repo=spritekindev/oss-user-6
2023/09/01 12:29:19 INFO created workspace id=ws-ipRXmQVDr94eJQ1r name=oss-user-6-oss-prod-changedetection-36 organization=Spritekin subject=Spritekin
```

First line you can see I deleted a workspace named oss-user-6-oss-prod-changedetection-35. I immediately got a list of error messages. Then I created a new workspace called oss-user-6-oss-prod-changedetection-36 but as you can see in the logs this workspace is not printing or starting the runs. However I do have enqueued the runs but they are locked in a pending state. 

<img width="895" alt="Screenshot 2023-09-01 at 10 38 38 pm" src="https://github.com/leg100/otf/assets/13467740/cf67bcee-e446-4f52-a8c2-f0e1058b6698">

However if I restart the container (kubectl delete pod) then the run appears in the logs and runs:
```
% kubectl delete pod otf-otf-6655468cc8-6ssck
pod "otf-otf-6655468cc8-6ssck" deleted
% kubectl logs otf-otf-6655468cc8-cvpp5    <<<< new pod
2023/09/01 12:42:30 INFO started cache max_size=0 ttl=10m0s
2023/09/01 12:42:30 INFO connected to database connstr="postgres://otf-otf-postgresql:5432/otf?pool_max_conns=10"
2023/09/01 12:42:30 INFO goose: no migrations to run. current version: 20230806181842
2023/09/01 12:42:30 INFO set site admins admins=[] promoted=[] demoted=[]
2023/09/01 12:42:30 INFO set system hostname hostname=otf.oss.ossy.dev
2023/09/01 12:42:30 INFO started server address=[::]:8080 ssl=false
2023/09/01 12:42:30 INFO enqueued plan id=run-4aRewH2ADGcSFU9u subject=scheduler
2023/09/01 12:42:30 INFO enqueued plan id=run-a0FUY60VKx6HM30b subject=scheduler
2023/09/01 12:42:30 INFO started plan id=run-4aRewH2ADGcSFU9u subject=local-agent
2023/09/01 12:42:30 INFO started plan id=run-a0FUY60VKx6HM30b subject=local-agent
2023/09/01 12:42:30 INFO executing phase component=agent run=run-4aRewH2ADGcSFU9u phase=plan
2023/09/01 12:42:30 INFO executing phase component=agent run=run-a0FUY60VKx6HM30b phase=plan
```
You can see the runs (from the screenshot) start immediately.

**Comment by leg100** [comment]
Ok, I've looked into this. The problem is the "scheduler", which is responsible for managing workspace queues and starting runs. The cause is this line:

```
2023/09/01 12:22:57 ERROR unsubscribing full subscriber component=broker error=<nil> sub=notifier-F1iB queue_length=100
```

The scheduler subscribes to events from the broker. However, the subscriber queue is limited in size and sometimes it gets full. That is happening for you because the workspace is deleted (one event), and so are all its runs (lots of events). When that happens the broker unsubscribes the subscriber, which in this case is the scheduler. The scheduler should handle this and re-subscribe. But I suspect it is not. Instead it is silently shutting down. And so any runs thereafter are stuck in the pending state.

I won't fix this just yet. Instead I've added some logging. I've also silenced the distracting "workspace queue does not exist for run event" log messages, which are just noise.

After I push v0.1.9, please try again, and see if the scheduler is indeed shutting down. You should see:

```
2023/09/02 11:39:37 DEBUG gracefully shutdown subsystem name=scheduler
```

**Comment by Spritekin** [comment]
Thanks... I was looking at the code just now just trying to understand it, as I mentioned Go is not my language of choice.
Sure I will upgrade and try again as soon as the new version is up.

**Comment by Spritekin** [comment]
@leg100 
I installed v0.1.9 and did some tests. Here are the new logs:
```
2023/09/04 12:24:19 INFO deleted workspace id=ws-vtUNeSTf9coONUa6 name=oss-user-6-oss-prod-changedetection-37 subject=Spritekin
2023/09/04 12:24:19 ERROR unsubscribing full subscriber component=broker error=<nil> sub=chunk-proxy-d501 queue_length=100
2023/09/04 12:24:19 ERROR unsubscribing full subscriber component=broker error=<nil> sub=purger-z5sE queue_length=100
2023/09/04 12:24:19 ERROR unsubscribing full subscriber component=broker error=<nil> sub=run-watch-d8JI queue_length=100
2023/09/04 12:24:19 ERROR unsubscribing full subscriber component=broker error=<nil> sub=notifier-HfXw queue_length=100
2023/09/04 12:24:19 ERROR unsubscribing full subscriber component=broker error=<nil> sub=scheduler-IzPB queue_length=100
2023/09/04 12:24:19 ERROR unsubscribing full subscriber component=broker error=<nil> sub=reporter-tf2y queue_length=100
2023/09/04 12:24:19 ERROR unsubscribing full subscriber component=broker error=<nil> sub=run-watch-JzYz queue_length=100
2023/09/04 12:31:22 INFO updated webhook webhook.id=ca33e812-4b94-4269-938a-a7e738edad9a webhook.vcs_provider_id=vcs-C05aynPJVWi9MdC4 webhook.cloud=github webhook.repo=spritekindev/oss-user-6 webhook.endpoint=https://otf.oss.ossy.dev/webhooks/vcs/ca33e812-4b94-4269-938a-a7e738edad9a webhook.vcs_id=430744892
2023/09/04 12:31:22 INFO connected workspace repo workspace=ws-3FJnfJuoNX6o8bJc subject=Spritekin repo=spritekindev/oss-user-6
2023/09/04 12:31:22 INFO created workspace id=ws-3FJnfJuoNX6o8bJc name=oss-user-6-oss-prod-changedetection-38 organization=Spritekin subject=Spritekin
```
The first line on the log is where I delete a workspace... I immediately get those messages.

Note all works OK until I delete a workspace, then the errors fire and no more runs are started.

**Comment by leg100** [comment]
Can you run it again, this time with `-v 9`?

**Comment by Spritekin** [comment]
@leg100 
Hi, I updated the env var to OTF_V=9 and I retried my tests. I had to create run and destroy workspaces a couple times before I found the error. This is the full log since the restart.
[logs.txt](https://github.com/leg100/otf/files/12523959/logs.txt)

Important time points:
2023/09/05 12:33:07 -> I destroyed a workspace then I got the errors.
2023/09/05 12:57:39 -> Created a new  workspace
2023/09/05 12:57:55 -> Created a new run on that workspace

You will notice from that point on there are no runs picked for execution, however I got runs pending but never started.

**Comment by Spritekin** [comment]
@leg100 
I have run tests and the new version does not fix the problem. Should I create a new item or reopen this one?

**Comment by leg100** [comment]
> @leg100 I have run tests and the new version does not fix the problem. Should I create a new item or reopen this one?

No, I'll have a look at it now.

**Comment by Spritekin** [comment]
@leg100 
I have tested this creating and destroying workspaces multiple times, plus installing and uninstalling resources. I have hit it even with 3 or 4 simultaneous requests. I haven't been able to make it crash so I guess I can consider this fixed.
